### DISCUSSION TITLE
Distinzione tra Fatture OSS e ExtraCEE

### DISCUSSION THREAD

**Discussion by dmigliacci** [fattureincloud/api#462] [Q&A] [answered]
Buongiorno,
in fase di creazione delle fatture, c'è un campo, flag, qualcosa che distingua le fatture destinate al [Sistema OSS](https://www.agenziaentrate.gov.it/portale/web/guest/risposte-alle-domande-piu-frequenti-oss-imprese) dalle altre fatture ExtraEU?

Grazie in anticipo

**Comment by valmoz** [comment]
Buongiorno @dmigliacci 
le consiglio di guardare questa guida: https://help.fattureincloud.it/help/articolo/653-operazioni-oss-ioss

Ci faccia sapere se le è sufficiente per risolvere il problema.

Grazie
Mauro

**Reply by dmigliacci** [reply]
Buongiorno @valmoz, grazie intanto per la risposta.
Mi sembra che l'argomento sia proprio quello corretto, mi leggo bene la guida, ma per chiarezza la mia necessità è che tale impostazione (vedi attributi avanzati) sia possibile farla tramite API.

Cioè creando una nuova fattura tramite API io posso impostare questa particolare configurazione.

Questo è possibile?

**Reply by dmigliacci** [reply]
Da quel che ho capito, in realtà, basta impostare per le aliquote OSS la loro Natura che poi sarà ereditata dalle fatture create tramite API. Corretto?

> Dal 1° luglio 2021 sono diventati operativi i nuovi regimi fiscali agevolati (One Stop Shop – Oss e Import One Stop Shop – Ioss) per le vendite a distanza di beni destinati a consumatori finali di altro Stato membro.
> 
> Se aderisci al regime agevolato puoi avvalerti della fatturazione elettronica per documentare le operazioni.
> 
> Il codice Natura IVA da utilizzare per esporre l'imponibile della vendita può essere:
> 
> IVA 0% Natura N3.2 non imponibili – cessioni intracomunitarie se si intende conteggiare l’operazione di vendita in regime OSS nel plafond IVA.
> IVA 0% Natura N7 - IVA assolta in altro stato UE qualora non si intende far concorrere l'operazione alla formazione del planford o per il computo delle richieste di rimborso trimestrale IVA.
> 
> Per quanto riguarda l’esposizione dell’imposta, oltre a poter utilizzare la sezione AltriDatiGestionali, è possibile utilizzare un'altra riga della fattura, esponendo l’importo IVA nel campo imponibile e utilizzando:
> 
> IVA 0% Natura N2.2 non soggette – altri casi oppure
> IVA 0% Natura N1 - Escluse ex art.15

![Screenshot 2024-07-23 alle 09 19 25](https://github.com/user-attachments/assets/50a2c795-d81d-4fa2-a4a3-6526591af601)

**Reply by valmoz** [reply, marked as answer]
Buongiorno @dmigliacci 
purtroppo non sono un esperto a riguardo di questo argomento, nel caso in cui dovessero servirle maggiori dettagli mi trovo costretto a rimandarla al Supporto Clienti.

Per quanto riguarda l'utilizzo lato API, leggendo la guida sembrerebbe essere sufficiente selezionare il VAT Type tra quelli preconfigurati, oppure crearne uno e poi usarlo esattamente come i vat types preconfigurati. 
Inoltre in alcuni casi sembrerebbe necessario utilizzare gli attributi avanzati, che lato API sono rappresentati tramite il campo ei_raw.

Secondo me per toglierci ogni dubbio potrebbe essere utile creare una fattura dal sito Fatture in Cloud con le caratteristiche necessarie, e poi recuperare la fattura creata tramite le API in modo tale da poter vedere il body. Il risultato della GET (o della LIST) sarà identico a quello che dovremo usare nella POST per creare una fattura simile, al netto degli ID del documento che non possono essere conosciuti a priori.

Di seguito le allego alcune guide che potrebbero esserle utili:
- [Invoice Creation](https://developers.fattureincloud.it/docs/guides/invoice-creation)
- [Invoice XML Customization](https://developers.fattureincloud.it/docs/guides/e-invoice-xml-customisation) - per gli attributi avanzati
- [List Issued Documents](https://developers.fattureincloud.it/api-reference#get-/c/-company_id-/issued_documents) e [Get Issued Document](https://developers.fattureincloud.it/api-reference#get-/c/-company_id-/issued_documents/-document_id-)

Grazie
Mauro

**Reply by dmigliacci** [reply]
Grazie Mauro,
spiegazione super completa!

Ho seguito la guida per impostare un parametro e seppur tutto sembra corretto, poi non mi permette più di inviare la fattura.

![Screenshot 2024-07-24 alle 15 44 49](https://github.com/user-attachments/assets/be80123a-6061-430e-aa6d-e691ac6b4824)

Nella dicitura del documento, sembra impostato come dico io

![Screenshot 2024-07-24 alle 15 47 12](https://github.com/user-attachments/assets/ac9ec735-7a69-45a9-9e4b-90581f91621c)

ma poi dentro la preview non compare ...

![Screenshot 2024-07-24 alle 15 47 23](https://github.com/user-attachments/assets/a301e04e-2c71-41e6-87ea-16a2f04aece5)

**Reply by emajo** [reply]
Ciao @dmigliacci quei problemi sono dovuti sicuramente a un settaggio incorretto del campo ei_raw, può mandarci il contenuto di quel campo che sta passando in creazione del documento?

Grazie,
Emanuele

**Reply by dmigliacci** [reply]
@emajo purtroppo avendo java 1.8 non posso usare l'esempio con Map.of

Il mio codice al momento sarebbe questo:

```
Map<String, String> tipoDocumentoMap = new HashMap<>();
tipoDocumentoMap.put("TipoDocumento", "TD01");
List<Map<String, String>> datiGeneraliDocumentoList = new ArrayList<>();
datiGeneraliDocumentoList.add(tipoDocumentoMap);
Map<String, List<Map<String, String>>> datiGeneraliDocumentoMap = new HashMap<>();
datiGeneraliDocumentoMap.put("DatiGeneraliDocumento", datiGeneraliDocumentoList);
Map<String, Map<String, List<Map<String, String>>>> datiGeneraliMap = new HashMap<>();
datiGeneraliMap.put("DatiGenerali", datiGeneraliDocumentoMap);
Map<String, Map<String, Map<String, List<Map<String, String>>>>> eiRaw = new HashMap<>();
eiRaw.put("FatturaElettronicaBody", datiGeneraliMap);
document.eiRaw(eiRaw);
```

**Reply by valmoz** [reply]
Buongiorno @dmigliacci, credo che per ottenere il JSON le basterebbe fare il marshall di eiRaw con una qualsiasi libreria per la gestione dei JSON.

Comunque, se non sbaglio, il campo ei_raw dovrebbe essere il seguente:

```json
"ei_raw": {
  "FatturaElettronicaBody": {
    "DatiGenerali": {
      "DatiGeneraliDocumento":{
          "TipoDocumento": "TD01"
        }
    }
  }
}
```
Quindi penso che ci siano delle list di troppo nel suo codice.

In ogni caso, per indagini più approfondite sarebbe meglio accedere ai suoi log, ma per farlo abbiamo la necessità di ottenere il Company ID che non può essere condiviso su questa discussion pubblica.

Le consigliamo quindi di contattare il supporto clienti, allegando il link a questa discussione, in modo da poter ottenere il suo Company ID e procedere con analisi più approfondite.
La avviso che il supporto clienti purtroppo è soggetto ad un tempo di risposta più alto rispetto a quello di questa community.

Grazie 
Mauro

**Reply by dmigliacci** [reply]
Si avevo seguito la struttura sbagliata dell'esempio, mi sa che bastano solo hashmap (non ho voglia per ora di fare il marshall finchè il commercialista non mi dice tutti i campi che vuole davvero gestire)

```
Map<String, String> tipoDocumentoMap = new HashMap<>();
tipoDocumentoMap.put("TipoDocumento", "TD01");

Map<String, Object> datiGeneraliDocumentoMap = new HashMap<>();
datiGeneraliDocumentoMap.put("DatiGeneraliDocumento", tipoDocumentoMap);

Map<String, Object> datiGeneraliMap = new HashMap<>();
datiGeneraliMap.put("DatiGenerali", datiGeneraliDocumentoMap);

Map<String, Object> fatturaElettronicaBodyMap = new HashMap<>();
fatturaElettronicaBodyMap.put("FatturaElettronicaBody", datiGeneraliMap);

document.eiRaw(fatturaElettronicaBodyMap);
```

**Reply by dmigliacci** [reply]
<img width="814" alt="Screenshot 2024-07-26 alle 07 40 07" src="https://github.com/user-attachments/assets/6b44044f-dc04-44f7-b4a3-93a36e0b9d64">
Like a charme, ok grazie a tutti.

Credo sia stata una discussione molto utile
